zanata-adapter-xliff/src/main/java/org/zanata/adapter/xliff/XliffWriter.java

```diff
@@ -175,48 +189,56 @@ private static void writeTransUnitContext(IndentingXMLStreamWriter writer,
+    }
+
+    private static void writeFile(File file, Resource doc, String locale,
+        TranslationsResource targetDoc, boolean createSkeletons) {
+
+        FileOutputStream fileStream = null;
+        XMLStreamWriter xmlStreamWriter = null;
+        IndentingXMLStreamWriter writer = null;
+
         try {
             XMLOutputFactory output = XMLOutputFactory.newInstance();
-            File outFile =
-                    new File(baseDir, doc.getName() + "_"
-                            + locale.replace('-', '_') + ".xml");
-            PathUtil.makeParents(outFile);
-            FileOutputStream fileStream = new FileOutputStream(outFile);
-            try {
-                XMLStreamWriter xmlStreamWriter =
-                        output.createXMLStreamWriter(fileStream, "utf-8");
-                try {
-                    IndentingXMLStreamWriter writer =
-                            new IndentingXMLStreamWriter(xmlStreamWriter);
-                    try {
-                        if (targetDoc != null) {
-                            writeHeader(writer, doc, locale);
-                        } else {
-                            writeHeader(writer, doc, null);
-                        }
-                        writeTransUnits(writer, doc, targetDoc, createSkeletons);
-
-                        // end body tag
-                        writer.writeEndElement();
-                        // end file tag
-                        writer.writeEndElement();
-                        // end xliff tag
-                        writer.writeEndDocument();
-                        writer.flush();
-                    } finally {
-                        writer.close();
-                    }
-                } finally {
-                    xmlStreamWriter.close();
-                }
-            } finally {
-                fileStream.close();
-            }
+            PathUtil.makeParents(file);
+            fileStream = new FileOutputStream(file);
+
+            xmlStreamWriter = output.createXMLStreamWriter(fileStream, "utf-8");
+
+            writer = new IndentingXMLStreamWriter(xmlStreamWriter);
+            writeHeader(writer, doc, locale);
+
+            writeTransUnits(writer, doc, targetDoc, createSkeletons);
+
+            // end body tag
+            writer.writeEndElement();
+            // end file tag
+            writer.writeEndElement();
+            // end xliff tag
+            writer.writeEndDocument();
+            writer.flush();
         } catch (XMLStreamException e) {
             throw new RuntimeException("Error generating XLIFF file format   ",
-                    e);
+                e);
         } catch (IOException e) {
             throw new RuntimeException("Error writing XLIFF file  ", e);
+        } finally {
+            try {
+                if (writer != null) {
+                    writer.close();
+                }
+                xmlStreamWriter.close();
+                fileStream.close();
```

Comment:
Please see if you can use the try-with-resources pattern here, as this can leave `xmlStreamWriter` and `fileStream` open if there is an exception while trying to close writer.

See this card from a few retrospectives ago: https://trello.com/c/gj0btrCU/4-try-with-resources